### PR TITLE
Allow Hash::remove() to remove from ArrayAccess objects

### PR DESCRIPTION
**Pull request by othercorey**
closes https://github.com/cakephp/cakephp/issues/17542

### CODE REVIEW

**Comment by ADmad** [comment]
@othercorey Didn't my update get this working as expected?

**Comment by othercorey** [comment]
Oh, maybe? I didn't notice you pushed. I made similar changes and kept running into static analysis errors due to remove() accepting ArrayAccess as a parameter.

**Comment by ADmad** [comment]
My bad, I should have left a comment after updating. I'll look into the static analysis errors.

File: src/Utility/Hash.php [hunk 1]
```diff
@@ -336,11 +337,13 @@ public static function insert(array $data, string $path, mixed $values = null):
...
      * Perform a simple insert/remove operation.
      *
+     * @template T of \ArrayAccess|array
      * @param string $op The operation to do.
-     * @param \ArrayAccess|array $data The data to operate on.
+     * @param T $data The data to operate on.
      * @param list<string> $path The path to work on.
      * @param mixed $values The values to insert when doing inserts.
      * @return \ArrayAccess|array data.
+     * @psalm-return (T is array ? array : \ArrayAccess)
```
Comment (othercorey) [inline src/Utility/Hash.php]: I think this is what I was missing. I was expecting T to evaluate to a specific type in the above union.

File: src/Utility/Hash.php [hunk 2]
```diff
@@ -415,6 +420,10 @@ public static function remove(ArrayAccess|array $data, string $path): ArrayAcces
 
         [$token, $conditions] = self::_splitConditions($token);
 
+        if (!is_iterable($data)) {
+            throw new CakeException('Cannot use path tokens of type `{}` or `[]` for non-iterable objects.');
+        }
```
Comment (othercorey) [inline src/Utility/Hash.php]: Ok, this is why I closed the PR originally, I wasn't sure if we had to add ArrayAccess-only iteration and logic.

**Comment by othercorey** [comment]
I think this change requires changing insert() to match.

**Comment by othercorey** [comment]
@ADmad Added support to insert()

**Comment by ADmad** [comment]
LGTM. The psalm errors can be ignored, the iterable check seems to confuse it.

**Pull request merged**